### PR TITLE
Calculation changes

### PR DESCRIPTION
**Pull request by mattalway**
Not suggesting this is definitely how we should calculate the metrics, but it does provide an interesting comparison. Using an example project, this moved the file complexity from 26.94 to 24.71, so it does have a noticeable effect.

Coverage / Compliance unchanged at the moment - this is just a demo.

### CODE REVIEW

**Comment by adamrodger** [comment]
Sorry I've not had chance to review this yet, need to find the time!

File: scripts/sonar/sonar-controller.js [hunk 1]
```diff
@@ -41,23 +50,51 @@ angular.module('app.sonar').controller('SonarCtrl', ['$scope', '$log', '$q', 'ap
...
 
                 var weightedMetrics = timePeriod.metrics.map(function(project) {
-                    var factor = project.lines / totalLines;
 
                     return {
-                        methodComplexity: factor * project.methodComplexity,
-                        fileComplexity: factor * project.fileComplexity,
-                        coverage: factor * project.coverage,
-                        compliance: factor * project.compliance,
+                        numberOfMethodsInProjectInPeriod: project.totalMethods,
```
Comment (adamrodger) [inline scripts/sonar/sonar-controller.js]: I'm all for descriptive variable names, but these ones are probably a bit too wordy. Without 'InProjectInPeriod' I think they'd be fine, since that's implied from the context of the rest of the code that these are per project per time period.

File: scripts/sonar/sonar-controller.js [hunk 2]
```diff
@@ -41,23 +50,51 @@ angular.module('app.sonar').controller('SonarCtrl', ['$scope', '$log', '$q', 'ap
...
+                        numberOfFilesInProjectInPeriod: project.totalFiles,
+                        complexityInProjectInPeriod: project.totalComplexity,
+                        testableLinesInProjectInPeriod: project.testableLines,
+                        untestedLinesInProjectInPeriod: project.untestedLines,
+
```
Comment (adamrodger) [inline scripts/sonar/sonar-controller.js]: Probably best not to have a blank line in the middle of an object initialisation, it looks a bit odd.

File: scripts/sonar/sonar-controller.js [hunk 3]
```diff
@@ -41,23 +50,51 @@ angular.module('app.sonar').controller('SonarCtrl', ['$scope', '$log', '$q', 'ap
...
+                        weightedCriticalIssuesInProjectInPeriod: weighting.critical * project.criticalIssues,
+                        weightedMajorIssuesInProjectInPeriod: weighting.major * project.majorIssues,
+                        weightedMinorIssuesInProjectInPeriod: weighting.minor * project.minorIssues,
+                        weightedInfoIssuesInProjectInPeriod: weighting.info * project.infoIssues
+
                     };
                 });
-
+                
+                // Complexity calculations
```
Comment (adamrodger) [inline scripts/sonar/sonar-controller.js:60]: The calculations would probably be nice in 2 functions, one for complexity and one for compliance, just to cut down the length of this function a bit.

File: scripts/sonar/sonar-controller.js [hunk 4]
```diff
@@ -41,23 +50,51 @@ angular.module('app.sonar').controller('SonarCtrl', ['$scope', '$log', '$q', 'ap
...
+                                                + sum(weightedMetrics, 'weightedMajorIssuesInProjectInPeriod') + sum(weightedMetrics, 'weightedMinorIssuesInProjectInPeriod')
+                                                + sum(weightedMetrics, 'weightedInfoIssuesInProjectInPeriod');
+
+                var rulesComplianceInPeriod = (100 - (totalWeightedIssuesInPeriod / totalLines) * 100)
+                
+                // Coverage calculations
+                var testableLinesOfCodeInPeriod = sum(weightedMetrics, 'testableLinesInProjectInPeriod')
+                var untestedLinesOfCodeInPeriod = sum(weightedMetrics, 'untestedLinesInProjectInPeriod')
+                
+                var codeCoverageInPeriod = ((testableLinesOfCodeInPeriod - untestedLinesOfCodeInPeriod) / testableLinesOfCodeInPeriod) * 100
```
Comment (adamrodger) [inline scripts/sonar/sonar-controller.js]: Should be:

```
var codeCoverage = ((untestedLines / testableLines) * 100) || 0;
```

Then you don't need the `NaN` check either.

**Comment by adamrodger** [comment]
The maths makes a lot more sense than the previous way (which was copied over from how we used to calculate metrics).

A few code issues in there that I've added inline comments for.

Also, can you rebase onto the latest from the upstream repo and squash all the commits down into one please?

File: scripts/sonar/sonar-controller.js (hunk 4 above)
Comment (mattalway) [inline scripts/sonar/sonar-controller.js]: The guide so far has been to duplicate whatever logic Sonar has been using. This is the line_coverage metric as referenced below. I will implement the "|| 0" but the change to the actual calculation needs to be considered futher.

Ref: http://docs.sonarqube.org/display/SONAR/Metric+definitions

Comment (adamrodger) [inline scripts/sonar/sonar-controller.js]: Yep, my comment above there is off, that gives you the percentage of code uncovered, not covered. That should be subtracted from 100 to give a coverage percentage.

It could also be written as:

```
var codeCoverage = ((testableLines - uncoveredLines) / testableLines) * 100 || 0;
```

**Comment by adamrodger** [comment]
Not sure what's happened there with the commits, do you mind if I just squash them all down into one when I merge and mark you as author on the commit?

File: scripts/sonar/sonar-controller.js [hunk 5]
```diff
@@ -40,24 +41,40 @@ angular.module('app.sonar').controller('SonarCtrl', ['$scope', '$log', '$q', 'ap
...
                     return {
-                        methodComplexity: factor * project.methodComplexity,
-                        fileComplexity: factor * project.fileComplexity,
-                        coverage: factor * project.coverage,
-                        compliance: factor * project.compliance,
+                        numberOfMethods: project.totalMethods,
+                        numberOfFiles: project.totalFiles,
+                        projectComplexity: project.totalComplexity,
+                        projectTestableLines: project.testableLines,
+                        projectUntestableLines: project.untestedLines,
```
Comment (adamrodger) [inline scripts/sonar/sonar-controller.js]: Shouldn't this variable be called `untestedLines` instead of `untestableLines`?

**Comment by mattalway** [comment]
I don't know either. :) Yeah please go ahead with that. Cheers.

File: scripts/sonar/sonar-controller.js (hunk 5 above)
Comment (mattalway) [inline scripts/sonar/sonar-controller.js]: Yes it should be.

File: scripts/sonar/sonar-controller.js [hunk 6]
```diff
@@ -67,6 +84,31 @@ angular.module('app.sonar').controller('SonarCtrl', ['$scope', '$log', '$q', 'ap
...
+            
+            return (((testableLinesOfCode - untestedLinesOfCode) / testableLinesOfCode) * 100) || 0
+        }
+        
+        function calculateCompliance(sonarData, totalLines){
+            var totalWeightedIssues = sum(sonarData, 'weightedBlockingIssues') + sum(sonarData, 'weightedCriticalIssues')
+                                    + sum(sonarData, 'weightedMajorIssues') + sum(sonarData, 'weightedMinorIssues')
+                                    + sum(sonarData, 'weightedInfoIssues');
+                                    
+            return (100 - (totalWeightedIssues / totalLines) * 100)
```
Comment (adamrodger) [inline scripts/sonar/sonar-controller.js]: Your brackets need a tweak here, the SonarQube documentation says:

```
(RCI) = 100 - (Weighted issues / Lines of code * 100)
```

**Comment by adamrodger** [comment]
Merged in de14c03, thank you!
